lib/ace/ext/papeeria/katex-previewer.coffee

```diff
@@ -56,25 +56,103 @@ define((require, exports, module) ->
     getFormulaElement = -> $("#formula")
     KATEX_OPTIONS = {displayMode: true, throwOnError: false}
+    TokenIterator = require("ace/token_iterator").TokenIterator
+    Range = require("ace/range").Range
+
+    erh = EquationRangeHandler = {
+      BEGIN_EQUATION_TOKEN_SEQUENCE: [
+        {
+          type: "storage.type"
+          value: "\\begin"
+        }
+        {
+          type: "lparen"
+          value: "{"
+        }
+        {
+          type: "variable.parameter"
+          value: "equation"
+        }
+        {
+          type: "rparen"
+          value: "}"
+        }
+      ]
+      END_EQUATION_TOKEN_SEQUENCE: [
+        {
+          type: "storage.type"
+          value: "\\end"
+        }
+        {
+          type: "lparen"
+          value: "{"
+        }
+        {
+          type: "variable.parameter"
+          value: "equation"
+        }
+        {
+          type: "rparen"
+          value: "}"
+        }
+      ]
+      rangeCache: {}
+
+      compareTokens: (token1, token2) ->
+        token1["type"] == token2["type"] and token1["value"] == token2["value"]
+
+      getEquationStart: (tokenIterator) ->
+        j = erh.BEGIN_EQUATION_TOKEN_SEQUENCE.length - 1
+        curEquationStart = null
+        # in case we're actually inside `\begin{equation}`
+        for token in erh.BEGIN_EQUATION_TOKEN_SEQUENCE
+          if erh.compareTokens(token, tokenIterator.getCurrentToken())
+            tokenIterator.stepForward()
+        while j >= 0
+          if erh.compareTokens(erh.BEGIN_EQUATION_TOKEN_SEQUENCE[j], tokenIterator.stepBackward())
+            if j == erh.BEGIN_EQUATION_TOKEN_SEQUENCE.length - 1
+              curTokenPosition = tokenIterator.getCurrentTokenPosition()
+              curEquationStart = {
+                row: curTokenPosition.row
+                column: curTokenPosition.column + tokenIterator.getCurrentToken().value.length
+              }
+            j -= 1
+          else
+            j = erh.BEGIN_EQUATION_TOKEN_SEQUENCE.length - 1
+            curEquationStart = null
+        return curEquationStart
+
+      getEquationEnd: (tokenIterator) ->
+        j = 0
+        curEquationStart = null
+        while j < erh.END_EQUATION_TOKEN_SEQUENCE.length
+          if erh.compareTokens(erh.END_EQUATION_TOKEN_SEQUENCE[j], tokenIterator.stepForward())
+            if j == 0
+              curEquationStart = tokenIterator.getCurrentTokenPosition()
+            j += 1
+          else
+            j = 0
+            curEquationStart = null
+        return curEquationStart
+
+      getEquationRange: (row, column) ->
+        if not erh.rangeCache[[row, column]]
+          tokenIterator = new TokenIterator(editor.getSession(), row, column)
+          start = erh.getEquationStart(tokenIterator)
+          end = erh.getEquationEnd(tokenIterator)
+          range = new Range(start.row, start.column, end.row, end.column)
+          erh.rangeCache[[row, column]] = range
```

Comment:
it depends on its lifetime. If you create this map on every change in the code then no need to clean it, but if it lives longer and survives changes in the editor then it should be cleaned appropriately.

In fact, do you really need this cache? Search by exact (row,column) is not the best one anyway, should you move cursor one column left or right, and you have cache miss. Maybe just remove it now and in future consider piggybacking on Ace's anchors (or whatever they are called) which can move automatically when text changes?